app/Database/CourseInsertion.hs

```diff
@@ -39,10 +43,28 @@ saveGraphJSON jsonStr nameStr = do
                 insertMany_ $ map (\path -> path {pathGraph = gId}) paths
             return $ toResponse $ ("Success" :: String)
 
--- | Inserts course into the Courses table.
+
+-- | Check for duplicate key, then if unique, insert course into Courses table
+--safeInsertCourse :: Course -> ReaderT SqlBackend m (Maybe Course) -- this screws up 'insert'
+--safeInsertCourse :: Course -> ReaderT SqlBackend m () -- this screws up 'insert'
+safeInsertCourse course = do
+    maybeCourse <- selectFirst [CoursesCode ==. (name course)] [] -- maybeCourse :: [Entity Courses] ?
+    case maybeCourse of
+        Nothing -> insertCourse course
+        Just _ -> return () --'Nothing' doesn't work here.
```

Comment:
Again, you can look at the type of `insertCourse`. The last part is a `()` (it's like void or null in other languages), which means that the final value to be "returned" must be `()`.